File: php/whoisdomainmonitoring_sdk.php
<?php
declare(strict_types=1);

// WhoisDomainMonitoring SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

// Features record diagnostic state on the client as dynamic properties
// (_retry, _cache, _metrics, ...); allow them explicitly (PHP 8.2+
// deprecates implicit dynamic properties).
#[\AllowDynamicProperties]
class WhoisDomainMonitoringSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new WhoisDomainMonitoringUtility();
        $this->_utility = $utility;

        $config = WhoisDomainMonitoringConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features in the resolved order (make_options puts an explicit
        // list order first, else defaults to test-first). Ordering matters: the
        // `test` feature installs the base mock transport and the transport
        // features (retry/cache/netsim/proxy/ratelimit) wrap whatever is
        // current, so `test` must be added before them to sit at the base.
        $feature_opts = WhoisDomainMonitoringHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $featureorder = Struct::getpath($this->options, "__derived__.featureorder");
            if (is_array($featureorder)) {
                foreach ($featureorder as $fname) {
                    $fopts = WhoisDomainMonitoringHelpers::to_map($feature_opts[$fname] ?? null);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, WhoisDomainMonitoringFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return WhoisDomainMonitoringUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = WhoisDomainMonitoringHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = WhoisDomainMonitoringHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = WhoisDomainMonitoringHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new WhoisDomainMonitoringSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = WhoisDomainMonitoringHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = WhoisDomainMonitoringHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_dns_result = null;

    // Canonical facade: $client->DnsResult()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->dns_result()
    // resolves here too.
    public function DnsResult($data = null)
    {
        require_once __DIR__ . '/entity/dns_result_entity.php';
        if ($data === null) {
            if ($this->_dns_result === null) {
                $this->_dns_result = new DnsResultEntity($this, null);
            }
            return $this->_dns_result;
        }
        return new DnsResultEntity($this, $data);
    }


    private $_domain = null;

    // Canonical facade: $client->Domain()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->domain()
    // resolves here too.
    public function Domain($data = null)
    {
        require_once __DIR__ . '/entity/domain_entity.php';
        if ($data === null) {
            if ($this->_domain === null) {
                $this->_domain = new DomainEntity($this, null);
            }
            return $this->_domain;
        }
        return new DomainEntity($this, $data);
    }


    private $_email_validate = null;

    // Canonical facade: $client->EmailValidate()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->email_validate()
    // resolves here too.
    public function EmailValidate($data = null)
    {
        require_once __DIR__ . '/entity/email_validate_entity.php';
        if ($data === null) {
            if ($this->_email_validate === null) {
                $this->_email_validate = new EmailValidateEntity($this, null);
            }
            return $this->_email_validate;
        }
        return new EmailValidateEntity($this, $data);
    }


    private $_generate = null;

    // Canonical facade: $client->Generate()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->generate()
    // resolves here too.
    public function Generate($data = null)
    {
        require_once __DIR__ . '/entity/generate_entity.php';
        if ($data === null) {
            if ($this->_generate === null) {
                $this->_generate = new GenerateEntity($this, null);
            }
            return $this->_generate;
        }
        return new GenerateEntity($this, $data);
    }


    private $_grammar = null;

    // Canonical facade: $client->Grammar()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->grammar()
    // resolves here too.
    public function Grammar($data = null)
    {
        require_once __DIR__ . '/entity/grammar_entity.php';
        if ($data === null) {
            if ($this->_grammar === null) {
                $this->_grammar = new GrammarEntity($this, null);
            }
            return $this->_grammar;
        }
        return new GrammarEntity($this, $data);
    }


    private $_ipn = null;

    // Canonical facade: $client->Ipn()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->ipn()
    // resolves here too.
    public function Ipn($data = null)
    {
        require_once __DIR__ . '/entity/ipn_entity.php';
        if ($data === null) {
            if ($this->_ipn === null) {
                $this->_ipn = new IpnEntity($this, null);
            }
            return $this->_ipn;
        }
        return new IpnEntity($this, $data);
    }


    private $_redact = null;

    // Canonical facade: $client->Redact()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->redact()
    // resolves here too.
    public function Redact($data = null)
    {
        require_once __DIR__ . '/entity/redact_entity.php';
        if ($data === null) {
            if ($this->_redact === null) {
                $this->_redact = new RedactEntity($this, null);
            }
            return $this->_redact;
        }
        return new RedactEntity($this, $data);
    }


    private $_ssl = null;

    // Canonical facade: $client->Ssl()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->ssl()
    // resolves here too.
    public function Ssl($data = null)
    {
        require_once __DIR__ . '/entity/ssl_entity.php';
        if ($data === null) {
            if ($this->_ssl === null) {
                $this->_ssl = new SslEntity($this, null);
            }
            return $this->_ssl;
        }
        return new SslEntity($this, $data);
    }


    private $_utility = null;

    // Canonical facade: $client->Utility()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->utility()
    // resolves here too.
    public function Utility($data = null)
    {
        require_once __DIR__ . '/entity/utility_entity.php';
        if ($data === null) {
            if ($this->_utility === null) {
                $this->_utility = new UtilityEntity($this, null);
            }
            return $this->_utility;
        }
        return new UtilityEntity($this, $data);
    }


    private $_whoi = null;

    // Canonical facade: $client->Whoi()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->whoi()
    // resolves here too.
    public function Whoi($data = null)
    {
        require_once __DIR__ . '/entity/whoi_entity.php';
        if ($data === null) {
            if ($this->_whoi === null) {
                $this->_whoi = new WhoiEntity($this, null);
            }
            return $this->_whoi;
        }
        return new WhoiEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new WhoisDomainMonitoringSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}

File: py/whoisdomainmonitoring_sdk.py
# WhoisDomainMonitoring SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import WhoisDomainMonitoringUtility
from core.spec import WhoisDomainMonitoringSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import WhoisDomainMonitoringBaseFeature
from features import _make_feature


class WhoisDomainMonitoringSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = WhoisDomainMonitoringUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features in the resolved order (make_options puts an explicit
        # list order first, else defaults to test-first). Ordering matters: the
        # `test` feature installs the base mock transport and the transport
        # features (retry/cache/netsim/proxy/ratelimit) wrap whatever is
        # current, so `test` must be added before them to sit at the base.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            featureorder = vs.getpath(self.options, "__derived__.featureorder")
            if isinstance(featureorder, list):
                for fname in featureorder:
                    fopts = helpers.to_map(feature_opts.get(fname))
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return WhoisDomainMonitoringUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = WhoisDomainMonitoringSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def DnsResult(self, data=None) -> "DnsResultEntity":
        """Entity factory: client.DnsResult().list() / client.DnsResult().load({"id": ...})."""
        from entity.dns_result_entity import DnsResultEntity
        return DnsResultEntity(self, data)


    def Domain(self, data=None) -> "DomainEntity":
        """Entity factory: client.Domain().list() / client.Domain().load({"id": ...})."""
        from entity.domain_entity import DomainEntity
        return DomainEntity(self, data)


    def EmailValidate(self, data=None) -> "EmailValidateEntity":
        """Entity factory: client.EmailValidate().list() / client.EmailValidate().load({"id": ...})."""
        from entity.email_validate_entity import EmailValidateEntity
        return EmailValidateEntity(self, data)


    def Generate(self, data=None) -> "GenerateEntity":
        """Entity factory: client.Generate().list() / client.Generate().load({"id": ...})."""
        from entity.generate_entity import GenerateEntity
        return GenerateEntity(self, data)


    def Grammar(self, data=None) -> "GrammarEntity":
        """Entity factory: client.Grammar().list() / client.Grammar().load({"id": ...})."""
        from entity.grammar_entity import GrammarEntity
        return GrammarEntity(self, data)


    def Ipn(self, data=None) -> "IpnEntity":
        """Entity factory: client.Ipn().list() / client.Ipn().load({"id": ...})."""
        from entity.ipn_entity import IpnEntity
        return IpnEntity(self, data)


    def Redact(self, data=None) -> "RedactEntity":
        """Entity factory: client.Redact().list() / client.Redact().load({"id": ...})."""
        from entity.redact_entity import RedactEntity
        return RedactEntity(self, data)


    def Ssl(self, data=None) -> "SslEntity":
        """Entity factory: client.Ssl().list() / client.Ssl().load({"id": ...})."""
        from entity.ssl_entity import SslEntity
        return SslEntity(self, data)


    def Utility(self, data=None) -> "UtilityEntity":
        """Entity factory: client.Utility().list() / client.Utility().load({"id": ...})."""
        from entity.utility_entity import UtilityEntity
        return UtilityEntity(self, data)


    def Whoi(self, data=None) -> "WhoiEntity":
        """Entity factory: client.Whoi().list() / client.Whoi().load({"id": ...})."""
        from entity.whoi_entity import WhoiEntity
        return WhoiEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "WhoisDomainMonitoringSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.dns_result_entity import DnsResultEntity
    from entity.domain_entity import DomainEntity
    from entity.email_validate_entity import EmailValidateEntity
    from entity.generate_entity import GenerateEntity
    from entity.grammar_entity import GrammarEntity
    from entity.ipn_entity import IpnEntity
    from entity.redact_entity import RedactEntity
    from entity.ssl_entity import SslEntity
    from entity.utility_entity import UtilityEntity
    from entity.whoi_entity import WhoiEntity
